### PR TITLE
cpu: RTC implementation for STM32F1

### PR DESCRIPTION
**Pull request by Former**
Implement real time clock for stm32f1.

Tested on bluepill board. 
Use tests: /tests/periph_rtc and /tests/periph_pm
Alarm can wake up board with command "unblock_rtc 0 30" after command "off" or "set 0".
set_alarm work fine

### CODE REVIEW

File: cpu/stm32_common/periph/rtc.c [hunk 1]
```diff
@@ -322,4 +329,257 @@ void ISR_NAME(void)
...
+
+        // Second interrupt is enabled.
+        RTC->CRH |= RTC_CRH_SECIE;
+
+        // Enter configuration mode.
+        RTC->CRL |= RTC_CRL_CNF;
+
+        //If the input clock frequency (fRTCCLK) is 32.768 kHz, write 7FFFh in this register to get a signal period of 1 second.
+        RTC->PRLH = 0;
+        RTC->PRLL = 0x8000;         //тактирование от внешнего кварца
```
Comment (maribu) [inline cpu/stm32_common/periph/rtc.c]: Please use English for comments and only use c-style comments. E.g.

``` C
/* this is a  c style comment */
// This is c++ style
```

(I'm aware that C starting from C99 mandates support for c++ style comments, but RIOT's coding convention explicitly forbids using c++ style comments.)

File: boards/common/stm32f103c8/Makefile.features [hunk 2]
```diff
@@ -2,6 +2,7 @@
 FEATURES_PROVIDED += periph_adc
 FEATURES_PROVIDED += periph_i2c
 FEATURES_PROVIDED += periph_pwm
+FEATURES_PROVIDED += periph_rtc
```
Comment (maribu) [inline boards/common/stm32f103c8/Makefile.features]: Please split everything related to the blue pill into a separate commit

File: cpu/stm32_common/periph/rtc.c [hunk 3]
```diff
@@ -322,4 +329,257 @@ void ISR_NAME(void)
...
+        //LSE oscillator clock used as RTC clock10: LSI oscillator clock used as RTC clock
+        RCC->BDCR |= RCC_BDCR_RTCEN | RCC_BDCR_RTCSEL_LSE;
+
+        // Second interrupt is enabled.
+        RTC->CRH |= RTC_CRH_SECIE;
+
+        // Enter configuration mode.
+        RTC->CRL |= RTC_CRL_CNF;
+
+        //If the input clock frequency (fRTCCLK) is 32.768 kHz, write 7FFFh in this register to get a signal period of 1 second.
```
Comment (maribu) [inline cpu/stm32_common/periph/rtc.c]: How does this comment relate to your code? You're writing `0x8000` into `RTC->PRLL`, should that be `0x7fff` instead?

File: cpu/stm32_common/periph/rtc.c [hunk 4]
```diff
@@ -322,4 +329,257 @@ void ISR_NAME(void)
...
+{
+    DEBUG("[RTC rtc_init]\n");
+
+    // Enable APB1 clocks
+    RCC->APB1ENR |= RCC_APB1ENR_PWREN | RCC_APB1ENR_BKPEN;
+
+    // Disable backup domain write protection
+    PWR->CR |= PWR_CR_DBP;
+
+    if ((RCC->BDCR & RCC_BDCR_RTCEN) != RCC_BDCR_RTCEN) // if RTC clock disabled
```
Comment (maribu) [inline cpu/stm32_common/periph/rtc.c]: The idea here is to skip initialization if it already has happened, right?

Generally, RIOT drivers should not assume that hardware is in a defined/well behaving state. E.g. assume a watchdog timer triggered a reboot in the middle of a configuration change. In that case a full reconfiguration of that hardware is required during driver initialization.

Back to this PR: I think this implies that the state of the RTC can only be one of the following:

a) disabled
b) enabled and correctly configured

Lets say the RTC  is instead indeed enabled, but incorrectly configured. In that case the initialization would fail.

File: cpu/stm32_common/periph/rtc.c [hunk 5]
```diff
@@ -322,4 +329,257 @@ void ISR_NAME(void)
...
+    _rtc_enter_config_mode();
+    RTC->CNTH = (counter_val & 0xffff0000) >> 16; // CNT[31:16]
+    RTC->CNTL = counter_val & 0x0000ffff;         // CNT[15:0]
+    _rtc_exit_config_mode();
+}
+
+// (UnixTime = 00:00:00 01.01.1970 = JD0 = 2440588)
+#define JULIAN_DATE_BASE    2440588
+
+// Julian day number calculation - https://en.wikipedia.org/wiki/Julian_day
```
Comment (maribu) [inline cpu/stm32_common/periph/rtc.c]: Why should the RTC use the Julian calendar?

> The Gregorian calendar is the most widely used civil calendar in the world.

https://en.wikipedia.org/wiki/Gregorian_calendar

File: cpu/stm32_common/periph/rtc.c [hunk 6]
```diff
@@ -322,4 +329,257 @@ void ISR_NAME(void)
...
+static void _rtc_get_time_from_count(const uint32_t rtc_counter, struct tm *time)
```
Comment (maribu) [inline cpu/stm32_common/periph/rtc.c]: Please just use [`localtime()`](https://linux.die.net/man/3/localtime) instead

File: cpu/stm32_common/periph/rtc.c [hunk 7]
```diff
@@ -322,4 +329,257 @@ void ISR_NAME(void)
...
+    time->tm_min = min;
+    time->tm_sec = sec;
+    time->tm_wday = wday;
+
+    DEBUG("[RTC time_from_count] c=%lu; %04i-%02i-%02i %02i:%02i:%02i\n", rtc_counter,
+        time->tm_year + 1900, time->tm_mon + 1, time->tm_mday,
+        time->tm_hour, time->tm_min, time->tm_sec);
+}
+
+// Julian day number calculation - https://en.wikipedia.org/wiki/Julian_day
```
Comment (maribu) [inline cpu/stm32_common/periph/rtc.c]: Please use the Gregorian calendar. (See above.)

File: cpu/stm32_common/periph/rtc.c [hunk 8]
```diff
@@ -322,4 +329,257 @@ void ISR_NAME(void)
...
+uint32_t _rtc_get_count_from_time(struct tm *time)
```
Comment (maribu) [inline cpu/stm32_common/periph/rtc.c]: Please just use [`mktime()`](https://linux.die.net/man/3/mktime) instead

File: cpu/stm32_common/periph/rtc.c [hunk 9]
```diff
@@ -322,4 +329,257 @@ void ISR_NAME(void)
...
+    uint32_t time_count_val = rtc_get_counter_val();
+    _rtc_get_time_from_count(time_count_val, time);
+    return 0;
+}
+
+int rtc_set_alarm(struct tm *time, rtc_alarm_cb_t cb, void *arg)
+{
+    (void)cb;
+    (void)arg;
+    if (cb != NULL) {
```
Comment (maribu) [inline cpu/stm32_common/periph/rtc.c]: How about using `assrert()` here?

File: cpu/stm32_common/periph/rtc.c [hunk 10]
```diff
@@ -322,4 +329,257 @@ void ISR_NAME(void)
...
+}
+
+int rtc_get_time(struct tm *time)
+{
+    uint32_t time_count_val = rtc_get_counter_val();
+    _rtc_get_time_from_count(time_count_val, time);
+    return 0;
+}
+
+int rtc_set_alarm(struct tm *time, rtc_alarm_cb_t cb, void *arg)
```
Comment (maribu) [inline cpu/stm32_common/periph/rtc.c]: This cannot be working, as both `cb` and `arg` are never used. The idea is that on the alarm `cb(arg)` is called. To do that, I would expect that both the function pointer and the argument are stored somewhere.

Have a lot at `struct isr_ctx;`, which is used by the other code and should be used here as well. Also the interrupt service routine `ISRNAME()` (yes, it is really named like that in the code) needs to be executed when the alarm is triggered. I didn't checked into the datasheet/documentation, so I have no idea how the isr needs to be registered.

File: cpu/stm32_common/periph/rtc.c (hunk 3 above)
Comment (Former) [reply cpu/stm32_common/periph/rtc.c]: Yes, 0x7fff will be correct

File: cpu/stm32_common/periph/rtc.c (hunk 4 above)
Comment (Former) [reply cpu/stm32_common/periph/rtc.c]: After config RTC drop current time.
Now rtc_init save current time. Config run always.

File: cpu/stm32_common/periph/rtc.c (hunk 10 above)
Comment (Former) [reply cpu/stm32_common/periph/rtc.c]: I use RTC only for wake up after turn off. Not use callback.
Now callback work fine.

**Comment by maribu** [comment]
Oh, I just noticed that the commit messages do not have the format "<area>: <summary>". E.g. the could be something like these:

1. `cpu: RTC implementation for STM32F1`
2. `boards/bluepill: FEATURES_PROVIDED +=  periph_rtc`

File: cpu/stm32_common/periph/rtc.c [hunk 11]
```diff
@@ -322,4 +326,231 @@ void ISR_NAME(void)
...
+    BKP->RTCCR |= 0;
+
+    /* second interrupt is disabled. */
+    RTC->CRH &= ~RTC_CRH_SECIE;
+
+    _rtc_enter_config_mode();
+
+    /* if the input clock frequency (fRTCCLK) is 32.768 kHz, write 7FFFh in this register to get a signal period of 1 second. */
+    RTC->PRLH = 0;
+    RTC->PRLL = 0x7FFF;
```
Comment (maribu) [inline cpu/stm32_common/periph/rtc.c]: Btw: `0x8000` is 32.768. With 32.768 kHz clock frequency, it should be indeed `0x8000` here, shouldn't it? The comment above should then be fixed as well.

File: cpu/stm32_common/periph/rtc.c [hunk 12]
```diff
@@ -322,4 +326,231 @@ void ISR_NAME(void)
...
+{
+    _rtc_enter_config_mode();
+    RTC->CNTH = (counter_val & 0xffff0000) >> 16;
+    RTC->CNTL = counter_val & 0x0000ffff;
+    _rtc_exit_config_mode();
+}
+
+void rtc_init(void)
+{
+    /* save current time if RTC already works */
```
Comment (maribu) [inline cpu/stm32_common/periph/rtc.c]: `rtc_init()` will be called from `periph_init()` implemented in `drivers/periph_common/init.c`. I don't see any reason why it would get called more than once during startup. Therefore, I would suggest to skip the check if the RTC is already running and just configure it here

File: cpu/stm32_common/periph/rtc.c (hunk 11 above)
Comment (kaspar030) [reply cpu/stm32_common/periph/rtc.c]: The note is actually taken verbatim from the datasheet.

File: cpu/stm32_common/periph/rtc.c (hunk 12 above)
Comment (kaspar030) [reply cpu/stm32_common/periph/rtc.c]: Wakeup from standby would be a reason.

File: cpu/stm32_common/periph/rtc.c (hunk 11 above)
Comment (maribu) [reply cpu/stm32_common/periph/rtc.c]: Sure, it will start counting from zero, so 0x7fff is indeed correct

File: cpu/stm32_common/periph/rtc.c (hunk 12 above)
Comment (maribu) [reply cpu/stm32_common/periph/rtc.c]: Indeed

**Comment by maribu** [comment]
Tested on the blue pill using `examples/default`

- Setting/getting current time works as expected
- Setting an alarm in the future does work, the callback is executed at the correct point in time (message `The alarm rang` is displayed)
- Setting an alarm in the future and clearing it before it is triggered does work, the callback is not executed
- Setting an alarm in the past does work. The documentation is not clear if this should be an error or not. (`rtc_set_alarm()` should return `-2` for an invalid time passed as `struct tm`, but not what to do on a valid date in the past.)

File: cpu/stm32_common/periph/rtc.c [hunk 13]
```diff
@@ -322,4 +326,231 @@ void ISR_NAME(void)
...
+
+    /* restore current time after config */
+    if (is_rtc_enable) {
+        _rtc_set_time(cur_time);
+    }
+}
+
+int rtc_set_time(struct tm *time)
+{
+    time_t timestamp = mktime(time);
```
Comment (maribu) [inline cpu/stm32_common/periph/rtc.c]: `mktime()` returns `(time_t) -1` on error. In that case `rtc_set_time()` should return `-1` according to the API doc.

File: cpu/stm32_common/periph/rtc.c [hunk 14]
```diff
@@ -322,4 +326,231 @@ void ISR_NAME(void)
...
+        NVIC_EnableIRQ(RTC_IRQn);
+        DEBUG("%s enable RTC IRQ\n", __func__);
+    }
+
+    /* save callback and argument */
+    isr_ctx.cb = cb;
+    isr_ctx.arg = arg;
+
+    /* set wakeup time */
+    time_t timestamp = mktime(time);
```
Comment (maribu) [inline cpu/stm32_common/periph/rtc.c]: `mktime()` returns `(time_t) -1` on error. In that case `rtc_set_alarm()` should return `-2` according to the API doc.

Btw: I'm personally would favor to always check the arguments for being plausible (e.g. `time` and `cb` should not be `NULL`, `time` should be in the future).

File: cpu/stm32_common/periph/rtc.c [hunk 15]
```diff
@@ -322,4 +326,231 @@ void ISR_NAME(void)
     cortexm_isr_end();
 }
 
+#elif defined(CPU_FAM_STM32F1)
```
Comment (aabadie) [inline cpu/stm32_common/periph/rtc.c]: I haven't looked in detail if some parts could actually be shared with the existing implementation but if it's not the case, it would be better to have a F1 specific rtc.c file in cpu/stm32f1/periph/rtc.c. If it's possible to make the initial common driver more generic, then I would try to adapt it.

File: cpu/stm32_common/periph/rtc.c (hunk 14 above)
Comment (Former) [reply cpu/stm32_common/periph/rtc.c]: I think that `cd` may be `NULL`. When `rtc_set_alarm()` use for wake up after standby callback not need.
Alarm time save in standby mode. In alarm time in past may encode several bytes data. It is possible to save several bytes in the standby mode. Maybe it can be useful for processor who do not have EEPROM.

Comment (maribu) [reply cpu/stm32_common/periph/rtc.c]: > I think that `cd` may be NULL. When `rtc_set_alarm()` use for wake up after standby callback not need.

You're right. Also, `void isr_rtc(void)` does properly check if `cb` is `NULL` prior to calling, so there is no risk of calling a `NULL`-function pointer :-)

> Alarm time save in standby mode. In alarm time in past may encode several bytes data. It is possible to save several bytes in the standby mode. Maybe it can be useful for processor who do not have EEPROM.

Someone re-purposing an RTC alarm to store and retrieve data would very likely not use the RTC API. But no other RTC implementation in RIOT does check for alarm time in the past nor is it required by the API doc. Therefore it would indeed be more consistent to not check here is well.

Comment (kaspar030) [reply cpu/stm32_common/periph/rtc.c]: > In alarm time in past may encode several bytes data. It is possible to save several bytes in the standby mode. Maybe it can be useful for processor who do not have EEPROM.

AFAIK the stm32 have backup memory which is part of the RTC peripheral. Still this shouldn't be accessed through th ertc API, as @maribu stated.

Comment (kaspar030) [reply cpu/stm32_common/periph/rtc.c]: I had some code lying around for using the backup registers, please see #11369.

Comment (Former) [reply cpu/stm32_common/periph/rtc.c]: > AFAIK the stm32 have backup memory which is part of the RTC peripheral. Still this shouldn't be accessed through th ertc API, as @maribu stated.

STM32f103 do not have EEPROM and RTC->BKP0R. Use `rtc_set_alarm()` is not good hack, but this may be one of the last chances.

Comment (maribu) [reply cpu/stm32_common/periph/rtc.c]: > STM32f103 do not have EEPROM and RTC->BKP0R. Use rtc_set_alarm() is not good hack, but this may be one of the last chances.

There is nothing wrong with resorting to `RTC->BKP0R` for storing non-RTC data. But I strongly believe that the RTC-API is not to be used for that. Instead a different API could be provided to use that, if someone wants to do that. (Feel free to open a separate PR to just do that, if you have a use case for that. This PR should also handle that the register is not used simultaneously as RTC alarm and as storage - maybe as simple as having `rtc_set_alarm()` return `-1` if the RTC-storage-API is used.)

Comment (maribu) [reply cpu/stm32_common/periph/rtc.c]: IMHO that would best be integrated into https://github.com/RIOT-OS/RIOT/pull/11369 to have a consistent API. Maybe a pseudo module could be used to unlock using the RTC register for just that.

**Comment by fjmolinas** [comment]
@Former 

Since rtt.c and rtc.c both use the same underlying RTC in STM32F1 I think you should add:

```
FEATURES_CONFLICT += periph_rtc:periph_rtt
FEATURES_CONFLICT_MSG += "On the STM32F1, the RTC and RTT map to the same hardware peripheral."
```
under `cpu/stm32f1/Makefile.features`

**Comment by fjmolinas** [comment]
Tested on iotlab-m3, it works except in sleep modes where the rtc alarm FAILS to wake up the device.

**Comment by fjmolinas** [comment]
> Tested on bluepill board.
> Use tests: /tests/periph_rtc and /tests/periph_pm
> Alarm can wake up board with command "unblock_rtc 0 30" after command "off" or "set 0".
> set_alarm work fine

I'm able to wake up from stand-by:

```
2019-09-19 08:30:02,963 - INFO # help
2019-09-19 08:30:02,966 - INFO # Command              Description
2019-09-19 08:30:02,969 - INFO # ---------------------------------------
2019-09-19 08:30:02,972 - INFO # off                  turn off
2019-09-19 08:30:02,975 - INFO # reboot               reboot
2019-09-19 08:30:02,978 - INFO # block                block power mode
2019-09-19 08:30:02,981 - INFO # set                  set power mode
2019-09-19 08:30:02,985 - INFO # unblock              unblock power mode
2019-09-19 08:30:02,989 - INFO # unblock_rtc          temporary unblock power mode
> unblock 0
2019-09-19 08:30:06,854 - INFO #  unblock 0
2019-09-19 08:30:06,856 - INFO # Unblocking power mode 0.
> unblock 1
2019-09-19 08:30:09,006 - INFO #  unblock 1
2019-09-19 08:30:09,008 - INFO # Unblocking power mode 1.
> unblock_rtc 1 3
2019-09-19 08:30:14,429 - INFO #  unblock_rtc 1 3
2019-09-19 08:30:14,432 - INFO # Unblocking power mode 1 for 3 seconds.
> help2019-09-19 08:30:18,961 - INFO #  �main(): This is RIOT! (Version: 2019.10-devel-588-g3043a-pr-11258)
2019-09-19 08:30:18,966 - INFO # This application allows you to test the CPU power management.
2019-09-19 08:30:18,972 - INFO # The available power modes are 0 - 1. Lower-numbered power modes
2019-09-19 08:30:18,977 - INFO # save more power, but may require an event/interrupt to wake up
2019-09-19 08:30:18,980 - INFO # the CPU. Reset the CPU if needed.
```

I'm unable to wake up form stop-mode:

```
make: Entering directory '/builds/tmp/RIOT/tests/periph_pm'
/builds/tmp/RIOT/dist/tools/pyterm/pyterm -p "/dev/riot/tty-nucleo-f103rb" -b "115200" 
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2019-09-19 08:29:15,103 - INFO # Connect to serial port /dev/riot/tty-nucleo-f103rb
helpWelcome to pyterm!
Type '/exit' to exit.
help
2019-09-19 08:29:16,529 - INFO # help
2019-09-19 08:29:16,532 - INFO # Command              Description
2019-09-19 08:29:16,536 - INFO # ---------------------------------------
2019-09-19 08:29:16,538 - INFO # off                  turn off
2019-09-19 08:29:16,541 - INFO # reboot               reboot
2019-09-19 08:29:16,544 - INFO # block                block power mode
2019-09-19 08:29:16,548 - INFO # set                  set power mode
2019-09-19 08:29:16,551 - INFO # unblock              unblock power mode
2019-09-19 08:29:16,556 - INFO # unblock_rtc          temporary unblock power mode
> unblock 1
2019-09-19 08:29:21,751 - INFO #  unblock 1
2019-09-19 08:29:21,753 - INFO # Unblocking power mode 1.
> unblock_rtc 1 3
2019-09-19 08:29:26,963 - INFO #  unblock_rtc 1 3
2019-09-19 08:29:26,966 - INFO # Unblocking power mode 1 for 3 seconds.
> help
help
help
help
help
help
```

I didn't work for me when rebasing on #11286 either.

**Comment by fjmolinas** [comment]
Might it be missing the EXTI configuration?

**Comment by benpicco** [comment]
This needs a rebase.
I think this PR should be good, stop-mode might be a separate issue.

**Comment by Former** [comment]
I rebase branch and fix conflicts.

File: boards/common/arduino-atmega/Makefile.features [hunk 16]
```diff
@@ -2,6 +2,7 @@
 FEATURES_PROVIDED += periph_adc
 FEATURES_PROVIDED += periph_i2c
 FEATURES_PROVIDED += periph_pwm
+FEATURES_PROVIDED += periph_rtc
```
Comment (benpicco) [inline boards/common/arduino-atmega/Makefile.features]: This is unrelated.

File: cpu/stm32f1/periph/rtc.c [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+        NVIC_EnableIRQ(RTC_IRQn);
+        DEBUG("%s enable RTC IRQ\n", __func__);
+    }
+
+    /* save callback and argument */
+    isr_ctx.cb = cb;
+    isr_ctx.arg = arg;
+
+    /* set wakeup time */
+    time_t timestamp = mktime(time);
```
Comment (benpicco) [inline cpu/stm32f1/periph/rtc.c]: What will you do if 19 January 2038 rolls around?

File: boards/common/arduino-atmega/Makefile.features (hunk 16 above)
Comment (benpicco) [reply boards/common/arduino-atmega/Makefile.features]: Should go to `cpu/stm32f1/Makefile.features`.

**Comment by benpicco** [comment]
When I run `tests/periph_rtc` I only see one alarm:

```
2019-11-28 22:49:22,690 # main(): This is RIOT! (Version: 2020.01-devel-1083-g37dc6)
2019-11-28 22:49:22,691 # 
2019-11-28 22:49:22,693 # RIOT RTC low-level driver test
2019-11-28 22:49:22,698 # This test will display 'Alarm!' every 2 seconds for 4 times
2019-11-28 22:49:22,702 #   Setting clock to   2011-12-13 14:15:57
2019-11-28 22:49:22,705 # Clock value is now   2011-12-13 14:15:57
2019-11-28 22:49:22,709 #   Setting alarm to   2011-12-13 14:15:59
2019-11-28 22:49:22,713 #    Alarm is set to   2011-12-13 14:15:59
2019-11-28 22:49:22,714 # 
2019-11-28 22:49:25,703 # Alarm!
```

File: cpu/stm32f1/periph/rtc.c (hunk 17 above)
Comment (Former) [reply cpu/stm32f1/periph/rtc.c]: cpu_efm32 - same problem. IMHO it`s a problem mktime function. 
Your solution?

Comment (benpicco) [reply cpu/stm32f1/periph/rtc.c]: Yea we could use an offset, but that should be coordinated between all RTC implementations.

**Comment by benpicco** [comment]
Murdock found some [minor issues](https://ci.riot-os.org/RIOT-OS/RIOT/11258/c1fd3b84c5083da1d2b736e9af471469d55693a3/output/compile/examples/default/iotlab-m3:llvm.txt) but `tests/periph_rtc` is not working as expected.

**Comment by Former** [comment]
> Murdock found some [minor issues](https://ci.riot-os.org/RIOT-OS/RIOT/11258/c1fd3b84c5083da1d2b736e9af471469d55693a3/output/compile/examples/default/iotlab-m3:llvm.txt) but `tests/periph_rtc` is not working as expected.

Fix this on 5d5357b. Use PRIu32.
tests/periph_rtc fix in 7e3921f

File: cpu/stm32f1/periph/rtc.c [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+
+    DEBUG("%s timestamp=%"PRIu32"\n", __func__, (uint32_t)timestamp);
+
+    return 0;
+}
+
+void rtc_clear_alarm(void)
+{
+    _rtc_disable_alarm();
+    _rtc_set_alarm_time(0);
```
Comment (benpicco) [inline cpu/stm32f1/periph/rtc.c]: Is this necessary?

File: cpu/stm32f1/periph/rtc.c [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+
+void rtc_poweroff(void)
+{
+    /* RTC is always on */
+    return;
+}
+
+void isr_rtc(void)
+{
+	struct alarm_isr_ctx_t ctx = isr_ctx;
```
Comment (benpicco) [inline cpu/stm32f1/periph/rtc.c]: tab instead of 4 spaces 😉

File: cpu/stm32f1/periph/rtc.c (hunk 18 above)
Comment (Former) [reply cpu/stm32f1/periph/rtc.c]: Save _rtc_set_alarm_time(0) in rtc_clear_alarm(void), but not _rtc_set_alarm_time(0) in callback.

File: cpu/stm32f1/periph/rtc.c [hunk 20]
```diff
@@ -0,0 +1,267 @@
...
+    return;
+}
+
+void isr_rtc(void)
+{
+    struct alarm_isr_ctx_t ctx = isr_ctx;
+
+    _rtc_disable_alarm();
+    _rtc_clear_isr_ctx();
+    cortexm_isr_end();
```
Comment (benpicco) [inline cpu/stm32f1/periph/rtc.c]: This should be the last line of the function.

File: cpu/stm32f1/periph/rtc.c [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+    /* RTC is always on */
+    return;
+}
+
+void isr_rtc(void)
+{
+    struct alarm_isr_ctx_t ctx = isr_ctx;
+
+    _rtc_disable_alarm();
+    _rtc_clear_isr_ctx();
```
Comment (benpicco) [inline cpu/stm32f1/periph/rtc.c]: Huh? Why is this needed here?
Now the callback will never be called.

File: cpu/stm32f1/periph/rtc.c (hunk 20 above)
Comment (Former) [reply cpu/stm32f1/periph/rtc.c]: "_rtc_disable_alarm" and other must run before callback. In callback may run new alarm. If run callback first, "_rtc_disable_alarm" turn off callback alarm.

File: cpu/stm32f1/periph/rtc.c (hunk 21 above)
Comment (Former) [reply cpu/stm32f1/periph/rtc.c]: I check this code before commit. Callback called fine. See copy `ctx = isr_ctx`. In `ctx` saved copy callback data.

File: cpu/stm32f1/periph/rtc.c (hunk 20 above)
Comment (Former) [reply cpu/stm32f1/periph/rtc.c]: I check in tests/periph_rtc. Callback call 4 times.

Comment (benpicco) [reply cpu/stm32f1/periph/rtc.c]: I mean the `cortexm_isr_end()`

File: cpu/stm32f1/periph/rtc.c (hunk 21 above)
Comment (benpicco) [reply cpu/stm32f1/periph/rtc.c]: Oh, didn't notice.
But this seems entirely unnecessary. Why do you need to delete the callback?
Disabling the alarm ought to be enough.

File: cpu/stm32f1/periph/rtc.c [hunk 22]
```diff
@@ -0,0 +1,265 @@
...
+    isr_ctx.cb = NULL;
+    isr_ctx.arg = NULL;
+}
+
+void rtc_clear_alarm(void)
+{
+    _rtc_disable_alarm();
+    _rtc_set_alarm_time(0);
+
+    _rtc_clear_isr_ctx();
```
Comment (benpicco) [inline cpu/stm32f1/periph/rtc.c]: I think you don't need this here either.

Comment (Former) [reply cpu/stm32f1/periph/rtc.c]: `_rtc_clear_isr_ctx`?
On other processor efm32/rtc, esp32/rtc and other callback is cleared.
I think clear need.

Comment (benpicco) [reply cpu/stm32f1/periph/rtc.c]: just "because others are doing it" is never good reasoning 😉 

The alarm can only be enabled again by `rtc_set_alarm()` where those are overwritten anyway.
It does no harm, but it's a waste of instructions and adds needless clutter to the code.

Comment (benpicco) [reply cpu/stm32f1/periph/rtc.c]: You can drop the `_rtc_set_alarm_time(0)` too.

Comment (Former) [reply cpu/stm32f1/periph/rtc.c]: See other `rtc_clear_alarm(void)`. All contain callback clear.

Comment (Former) [reply cpu/stm32f1/periph/rtc.c]: > You can drop the `_rtc_set_alarm_time(0)` too.

Function name - `rtc_clear_alarm`, not `rtc_disable_alarm`.
After `rtc_clear_alarm` we can run `rtc_get_alarm` and see that alarm not clear - only disable

Comment (benpicco) [reply cpu/stm32f1/periph/rtc.c]: `lpc2387` and  `cc430` do not… 
But yea there seems to be some strange cargo cult going on.

Well keep it if you want.
But the `_rtc_set_alarm_time(0)` won't help you if the counter wraps around and `_rtc_disable_alarm()` was not working for some reason.

Comment (Former) [reply cpu/stm32f1/periph/rtc.c]: > `lpc2387` and `cc430` do not…
> But yea there seems to be some strange cargo cult going on.
> 
> Well keep it if you want.
> But the `_rtc_set_alarm_time(0)` won't help you if the counter wraps around and `_rtc_disable_alarm()` was not working for some reason. (Which would be the only reason why you would want to delete the callback)

I found case.
We set time to 15:20:59, and set alarm 15:21:10. Alarm called.
At cur time 15:21:20 - set time 15:20:59.
Now we have: time 15:20:59. Alarm 15:21:10 (`rtc_get_alarm` return). But alarm not called (is disable)
I think that after callback we need clear alarm time!

Comment (Former) [reply cpu/stm32f1/periph/rtc.c]: > just "because others are doing it" is never good reasoning wink
> 
> The alarm can only be enabled again by `rtc_set_alarm()` where those are overwritten anyway.
> It does no harm, but it's a waste of instructions and adds needless clutter to the code.

Ok. I can argee with `clear_isr_ctx`. `isr_ctx` not visible outside. But alarm time visible on `rtc_get_alarm`.

Comment (benpicco) [reply cpu/stm32f1/periph/rtc.c]: Hm I would have expected `rtc_get_alarm()` to return -1 if no alarm is set, but this doesn't seem to be the case.

So returning an alarm in the past might be the next best thing.
IMHO the RTC API needs some overhaul…

Comment (Former) [reply cpu/stm32f1/periph/rtc.c]: > Hm I would have expected `rtc_get_alarm()` to return -1 if no alarm is set, but this doesn't seem to be the case.
> 
> So returning an alarm in the past might be the next best thing.
> IMHO the RTC API needs some overhaul…

If clear alarm on callback, then test periph_rtc work wrong:
`    puts("Alarm!");`

`    if (++cnt < REPEAT) {`
`        struct tm time;`
`        rtc_get_alarm(&time);`
`        inc_secs(&time, PERIOD);`
`        rtc_set_alarm(&time, cb, NULL);`
`    }`
`rtc_get_alarm` - return zero time and next 3 alarm not run.
If `rtc_get_alarm` return `-1` when alarm is disable (on callback too) - same problem. Alarm run only 1 times. Returned value `rtc_get_alarm` not check in test. I think replace `rtc_get_alarm` to `rtc_get_time` - good things.

Comment (benpicco) [reply cpu/stm32f1/periph/rtc.c]: Yea like I said, I think the API needs a bit of rework to be less ambiguous.
Anyway, your implementation works and should not be help up over different interpretations of the unclear RTC API.
